### PR TITLE
ramboxpro: Migrate to using an AppImage

### PR DESCRIPTION
**Pull request by cawilliamson**
###### Motivation for this change
The original implementation of this package has always been rather hacky and is certainly not a supported way to run Rambox Pro. The problem only got worse when the recent migration away from Electron v5.X due to security concerns and the package has been in a broken state ever since (it builds but doesn't run.)

This package change rectifies both of these problems - it moves Rambox Pro to using a supported method of run (AppImage) and allows it to be easily upgraded going forward without the constant need to hack together a new solution for whatever changes the Rambox team make.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/rambox/pro.nix [hunk 1]
```diff
@@ -46,6 +13,27 @@ stdenv.mkDerivation rec {
...
 
+  src = fetchurl {
+    url = "https://github.com/ramboxapp/download/releases/download/v${version}/RamboxPro-${version}-linux-x64.AppImage";
+    name = "${name}.AppImage";
+    sha256 = "18383v3g26hd1czvw06gmjn8bdw2w9c7zb04zkfl6szgakrv26x4";
+  };
+
+  appimageContents = appimageTools.extract { inherit name src; };
+in
+{
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/rambox/pro.nix:24]: Please fix the eval error.

```suggestion
```

File: pkgs/applications/networking/instant-messengers/rambox/pro.nix [hunk 2]
```diff
@@ -46,6 +13,27 @@ stdenv.mkDerivation rec {
...
+  appimageTools.wrapType2 rec {
+    inherit name src;
+
+    extraInstallCommands = ''
+      mkdir -p $out/share/applications $out/share/icons/hicolor/256x256/apps
+
+      cp ${desktopItem}/share/applications/* $out/share/applications
+
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/rambox/pro.nix:32]:
```suggestion
      cp ${desktopItem}/share/applications/* $out/share/applications
```

File: pkgs/applications/networking/instant-messengers/rambox/pro.nix [hunk 3]
```diff
@@ -1,42 +1,9 @@
...
-in
-stdenv.mkDerivation rec {
-  pname = "rambox-pro";
+  name = "ramboxpro";
   version = "1.4.1";
 
-  dontBuild = true;
-  dontStrip = true;
-
-  buildInputs = [ nss xorg.libXext xorg.libxkbfile xorg.libXScrnSaver ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/rambox/pro.nix:13]: Why do we use the app image and not build it ourselves?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by thiagokokada** [comment]
> The original implementation of this package has always been rather hacky and is certainly not a supported way to run Rambox Pro. The problem only got worse when the recent migration away from Electron v5.X due to security concerns and the package has been in a broken state ever since (it builds but doesn't run.)

Are we going to migrate rambox-ce too :thinking: ?

**Comment by cawilliamson** [comment]
> Are we going to migrate rambox-ce too  ?

I would doubt it because the Rambox package is maintained by someone else and I don't believe it uses Electron so the root cause of most of the problems with the Pro variant aren't present for the free version.

File: pkgs/applications/networking/instant-messengers/rambox/pro.nix [hunk 4]
```diff
@@ -46,6 +13,23 @@ stdenv.mkDerivation rec {
...
+    inherit name src;
+
+    extraInstallCommands = ''
+      mkdir -p $out/share/applications $out/share/icons/hicolor/256x256/apps
+      cp ${desktopItem}/share/applications/* $out/share/applications
+      cp ${appimageContents}/usr/share/icons/hicolor/256x256/apps/ramboxpro.png $out/share/icons/hicolor/256x256/apps/ramboxpro.png
+    '';
+  };
+
   meta = with stdenv.lib; {
```
Comment (thiagokokada) [inline pkgs/applications/networking/instant-messengers/rambox/pro.nix:33]: `stdenv.lib` is deprecated. Please use `lib`.

**Review by thiagokokada — COMMENTED** [review]
I do think Rambox CE is a Electron app too, it is just that they repackaged the .deb instead (that I think includes a copy of Electron).

BTW, the reason I am asking this is because it seems both packages could have the exact same derivation, just with different src inputs. This would make maintaining both versions easier.

File: pkgs/applications/networking/instant-messengers/rambox/pro.nix [hunk 5]
```diff
@@ -46,6 +13,23 @@ stdenv.mkDerivation rec {
...
+  appimageContents = appimageTools.extract { inherit name src; };
+in appimageTools.wrapType2 rec {
+    inherit name src;
+
+    extraInstallCommands = ''
+      mkdir -p $out/share/applications $out/share/icons/hicolor/256x256/apps
+      cp ${desktopItem}/share/applications/* $out/share/applications
+      cp ${appimageContents}/usr/share/icons/hicolor/256x256/apps/ramboxpro.png $out/share/icons/hicolor/256x256/apps/ramboxpro.png
+    '';
+  };
```
Comment (thiagokokada) [inline pkgs/applications/networking/instant-messengers/rambox/pro.nix:31]: There still seems to be some eval issues, or GitHub syntax highlighting is wrong (but I think this very unlikely).

**Comment by thiagokokada** [comment]
I went and gave a try to converting both `rambox` and `rambox-pro` to use AppImage. This simplificated the derivation a lot: https://github.com/NixOS/nixpkgs/pull/109020

**Pull request closed without merging**